Add MQTT client authentication

Bravo555 opened the pull request:
## Proposed changes

This commit adds 2 new config options: `mqtt.client.auth.cert_file` and `mqtt.client.auth.key_file` to facilitate MQTT client authentication. The client certificate needs to be signed by the same CA as the server certificate for the authentication to work, that's why new settings were added instead of using a thin-edge device certificate located in /etc/tedge/device-certs, which is self-signed and adding the option of using CA to `tedge cert` command will not be trivial.

> Note: this PR is the second part of #1785 and requires server authentication to be present, which is why it contains commits from the server authentication PR (#1816). I anticipate #1816 to be merged before this PR which might need some more discussion and fixes, which is why I'm putting it as a draft. When #1816 is merged, the first 2 commits should disappear, leaving this PR with only commits related to client authentication.

## Types of changes

<!--
What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_
-->

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
- #1785

## Checklist

<!--
_Put an `x` in the boxes that apply. You can also fill these out after
creating the PR. If you're unsure about any of them, don't hesitate to ask.
We're here to help! This is simply a reminder of what we are going to look for
before merging your code._
-->

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

Some more thought needs to be put into how to integrate client authentication with device certificates generated by the `tedge cert create` command. The certificates generated by the command are currently self-signed, but for client authentication to work, both server and client need to use certificates signed by the same CA. A CA certificate is usually located on another device, to which the subject sends a CSR and receives back a signed certificate. So unless we want to keep it as it is now, where we use a certificate different from thin-edge device certificate, and which needs to be deployed by the user manually, the `tedge cert` command would have to be heavily modified.

didier-wenzek commented:
> The client certificate needs to be signed by the same CA as the server certificate for the authentication to work,

There is a misunderstanding here. This would be a too strong constraint. See [mosquitto documentation](https://mosquitto.org/man/mosquitto-conf-5.html): *This means that any CA certificates you include in cafile or capath will be able to issue client certificates that are valid for connecting to your broker.*

It means that to accept connections from some client, one must add in the mosquitto `capath` directory the signing certificate of this client certificate. Note that (still according mosquitto documentation):

         For capath to work correctly,
         the certificates files must have ".pem" as the file ending
         and you must run "openssl rehash <path to capath>" each time you add/remove a certificate.

> that's why new settings were added instead of using a thin-edge device certificate located in /etc/tedge/device-certs,

These two settings are required but for another reason. One needs to use different certificates and signing authority to authenticate the device in the cloud and to authenticate a local process or child-device.

Bravo555 commented:
> There is a misunderstanding here. This would be a too strong constraint. See [mosquitto documentation](https://mosquitto.org/man/mosquitto-conf-5.html): _This means that any CA certificates you include in cafile or capath will be able to issue client certificates that are valid for connecting to your broker._
> 
> It means that to accept connections from some client, one must add in the mosquitto `capath` directory the signing certificate of this client certificate. Note that (still according mosquitto documentation):

You're right, CAs trusted by the broker and the clients can actually be different.

> These two settings are required but for another reason. One needs to use different certificates and signing authority to authenticate the device in the cloud and to authenticate a local process or child-device.

Okay, then `tedge cert create` is probably not relevant. What prompted me to ask about it was that we have quite a lot of configuration options and I'm not sure whether or not I'm needlessly duplicating them instead of using some other options. E.g. for root trust store, e.g. I wondered instead of adding a key in `mqtt.client`, paths to trusted certificates should not be global, with some ability to override, e.g. with an options to add more paths or certificates in certain contexts, like HTTP or MQTT.

didier-wenzek commented on tests/RobotFramework/tests/mqtt/connect_broker.robot:
```diff
@@ -10,25 +10,45 @@ Force Tags      theme:mqtt
 Publish on a local insecure broker
     Start Service      mosquitto
-    Execute Command    tedge mqtt pub topic message
+    Execute Command    tedge mqtt pub topic1 message1
```
I would add a check that the message is actually received - here and for the two other tests using authentication

Bravo555 replied:
To verify that the message was received, I was thinking of starting `tedge mqtt pub` and `tedge mqtt sub` commands, but as `tedge sub` blocks, I don't really know how should I handle it using our ThinEdgeIO Robot Framework library. Could you give me some pointers? Because from the glimpse into the library, I don't see any way to start a background process on the test device.

reubenmiller replied:
@Bravo555 Sorry I'm a bit late...There is a mqtt logger which is running in the integration test device where you can use the `Should Have MQTT Messages` keyword. The mqtt logger is running all the time, so no need to explicitly do a mqtt sub.

Here is an example:

```
@{listen}=    Should Have MQTT Messages    topic=tedge/${CHILD_SN}/commands/req/config_update        date_from=-5s
```

didier-wenzek commented on crates/core/tedge/src/cli/mqtt/cli.rs:
```diff
@@ -61,6 +62,25 @@ impl BuildCommand for TEdgeMqttCli {
+            .load()?
+            .query(MqttClientAuthKeySetting);
+
+        let client_auth_config = if client_cert.is_err() && client_private_key.is_err() {
+            None
+        } else {
```
What the logic here? To ignore the config if both are incorrect?

Bravo555 replied:
Yes, if neither `client_cert` nor `client_key` are present, then we skip client authentication

didier-wenzek replied:
In see, one has to read `is_err()` as `is_none()`, but this is misleading as we have a `ConfigSettingError`. 

This is okay for now as there is another [on-going work on tedge-config](https://github.com/thin-edge/thin-edge.io/issues/1812). However, is see here yet another opportunity to improve  error management as you did with https://github.com/thin-edge/thin-edge.io/pull/1874: the same error type is used both for get and set.

didier-wenzek reviewed (commented):
That sounds correct. One point is missing though: this needs to be added to `mqtt_channel` too and used by the misc `tedge` daemons.

didier-wenzek commented:
> Okay, then tedge cert create is probably not relevant. What prompted me to ask about it was that we have quite a lot of configuration options and I'm not sure whether or not I'm needlessly duplicating them instead of using some other options. E.g. for root trust store, e.g. I wondered instead of adding a key in mqtt.client, paths to trusted certificates should not be global, with some ability to override, e.g. with an options to add more paths or certificates in certain contexts, like HTTP or MQTT.

Indeed, we start to have quite a lot of MQTT related options !

```
$ tedge config list --all | grep mqtt                                   
mqtt.bind_address=127.0.0.1
mqtt.client.host=localhost
mqtt.client.port=1883
mqtt.port=1883
mqtt.client.ca_file=
mqtt.client.ca_path=
mqtt.client.auth.cert_file=
mqtt.client.auth.key_file=
mqtt.external.port=
mqtt.external.bind_address=
mqtt.external.bind_interface=
mqtt.external.capath=
mqtt.external.certfile=
mqtt.external.keyfile=
```

The two added by this PR make sense and are related to the former ones
* `mqtt.external.*` are used to configure a listener on mosquitto.
* `mqtt.external.port` and `mqtt.client.port` should be the same for the client to connect that listener. Same for the addresses.
* `mqtt.external.capath` is where to put the signing certificate of a device certificate `mqtt.client.auth.cert_file`.
* `mqtt.client.ca_path` must contains a copy of the `mqtt.external.certfile`.


The more issue is more that `tedge_conf.toml` is mixing unrelated things: configuring mosquitto, configuring the tedge daemons and configuring tedge clients.

PS: Writing this comment, I noticed that something needs to be fixed: the client and external settings use the same names except an extra underscore (`capath` vs `ca_path`). @reubenmiller : what's your preference?

Bravo555 commented:
> One point is missing though: this needs to be added to `mqtt_channel` too and used by the misc `tedge` daemons.

I think there's one issue with `mqtt_channel` as it is now. Because it itself does not read thin-edge configuration, and uses a config builder, every extension using it has to read configuration and set common options, like hostname, port, or certificate paths. If I'm somewhat correct in thinking that `mqtt_channel` is supposed to be an abstraction which is supposed to be used by plugins to connect to thin-edge local broker and reduce some boilerplate resulting from all our plugins being MQTT clients, then it's not performing its task very well because when we add something that changes the behaviour of the broker, all the clients using `mqtt_channel` have to be updated as well. 

I think that instead of starting with `Config::default()`, which does not take config into account, we should provide a function that reads the thin-edge configuration and constructs `Config` from it. We could then allow plugins to override any fields they want, although It's not clear to me why they would even want to do so. All the information needed to connect to the broker is inside the config file, so I think there's no advantages to forcing clients to handle, or even be aware, of the transport protocol that is being used.

For the time being, I'll take the long road and just update all usages adding server and client authentication (in server authentication PR i updated `mqtt_channel`, but forgot to update usages). Please tell me if I'm understanding something wrong and whether or not the kind of changes that I propose above is something worth considering.

didier-wenzek reviewed (commented):
The changes on mqtt-channel to add client authentication support are correct.

The missing point is now to use this setting from all the tedge services. This might be an opportunity to add an `impl From<TEdgeConfig> for mqtt_channel::Config` as discussed here: https://github.com/thin-edge/thin-edge.io/pull/1864#issuecomment-1494500943. I will let you choose the steps you prefer to go there.

Bravo555 commented:
In [a48bdbc](https://github.com/thin-edge/thin-edge.io/pull/1864/commits/a48bdbca19a33d2c24d499e074f6755e646f2d0e), I did a first try in extracting common MQTT configuration options into one place. Made it just another impl function, because only a part of the configuration is extracted, and it involves calling builder methods on `mqtt_channel::Config`, so I think `From` wouldn't be the best fit. Made it in another module because it introduced a dependency on `mqtt_channel` in `tedge_config` so we can perhaps gate this module behind a feature flag if necessary.

Still, I'm not really satisfied with this, because by generating MQTT config from tedge config, now we need to pass `&TEdgeConfig` everywhere where we previously only needed `host` and `port`. This is particularly annoying in tests, and because constructing `TEdgeConfig` from scratch would be very painful, and because parsing configuration file in tests sounds like a bad idea, I've made it optional in some parts so it could be `None` in unit tests, but it's not supposed to be permanent. Now I'd like to try and run some tests that make use of all the affected components, and after I've confirmed that it works, I'll try to fix this "how to pass config" issue. My best guess would be to group `host`, `port` and `tls` into one object, but that would introduce nesting to `mqtt_channel`'s config builder which might be a bit awkward to use.

didier-wenzek commented:
> In [a48bdbc](https://github.com/thin-edge/thin-edge.io/pull/1864/commits/a48bdbca19a33d2c24d499e074f6755e646f2d0e),

This link is broken.

> I did a first try in extracting common MQTT configuration options into one place. Made it just another impl function, because only a part of the configuration is extracted, and it involves calling builder methods on `mqtt_channel::Config`, so I think `From` wouldn't be the best fit. Made it in another module because it introduced a dependency on `mqtt_channel` in `tedge_config` so we can perhaps gate this module behind a feature flag if necessary.

The function `TEdgeConfig::mqtt_config()` seems perfect.  Adding a feature flag might be not necessary as of today all the components have an MQTT dependency.
 
> Still, I'm not really satisfied with this, because by generating MQTT config from tedge config, now we need to pass `&TEdgeConfig` everywhere where we previously only needed `host` and `port`. This is particularly annoying in tests, and because constructing `TEdgeConfig` from scratch would be very painful, and because parsing configuration file in tests sounds like a bad idea, I've made it optional in some parts so it could be `None` in unit tests, but it's not supposed to be permanent. Now I'd like to try and run some tests that make use of all the affected components, and after I've confirmed that it works, I'll try to fix this "how to pass config" issue. My best guess would be to group `host`, `port` and `tls` into one object, but that would introduce nesting to `mqtt_channel`'s config builder which might be a bit awkward to use.

Here, there is something I don't understand. Why `&TEdgeConfig`  is now required where only  `host` and `port` were previously needed? Why not replace  `host` and `port`  parameters by an `mqtt_channel::Config` and let the caller use `TEdgeConfig::mqtt_config()` if required?

I agree that constructing `TEdgeConfig` from scratch is very painful but I would really avoid to make config parameter optional to ease testing. Test helpers can be added in [tedge_test_utils](https://github.com/thin-edge/thin-edge.io/tree/main/crates/tests/tedge_test_utils).

Also keep in mind that  [`TEdgeConfig` is under refactoring](https://github.com/thin-edge/thin-edge.io/issues/1812), and that it would be then feasible to write tests helpers as `TEdgeConfig::default().with_client_mqtt_port(1234)`.

PradeepKiruvale commented on tests/RobotFramework/tests/cumulocity/service_monitoring/service_monitoring.robot:
```diff
@@ -137,41 +139,41 @@ Check if a service is up
     Execute Command    tedge config set service.type thinedge
     Custom Test Setup
-    ThinEdgeIO.Start Service    ${service_name}
+    ThinEdgeIO.Restart Service    ${service_name}
```
Why restart is needed here?, is it already started by the `setup`?

Bravo555 replied:
The service starts when the connection to Cumulocity is not yet established, and fails with `c8y_api::http_proxy: An error occurred while retrieving internal Id, operation will retry in 60 seconds and mapper will reinitialize`. This error is mentioned earlier in the file, but bumping the timeout didn't really fix the problem. This did. If not having to restart the service is a hard requirement in this test, then I'll note it and be sure to find a solution that does not involve restarting, later.

didier-wenzek replied:
>  If not having to restart the service is a hard requirement in this test, then I'll note it and be sure to find a solution that does not involve restarting, later.

Restarting the service is okay.

PradeepKiruvale commented on crates/common/mqtt_channel/src/config.rs:
```diff
@@ -61,8 +64,37 @@ pub struct Config {
+///
+/// Intended to mirror authentication model found in the mosquitto MQTT broker,
+/// when certificate-based authentication is enabled, server authentication is
+/// mandatory and client authentication is optional.
```
The comment is not clear here, does this mean when the `server` authentication is enabled `clients` can connect without authentication?

Bravo555 replied:
there are 3 supported modes of connecting:

1. no authentication
2. server authentication - clients will verify MQTT broker certificate
3. server and client authentication - clients will verify MQTT broker certificate and broker will verify client certificates

Will reword the comment to make it more clear and include link to the relevant mosquitto.conf manual section

PradeepKiruvale commented on crates/common/mqtt_channel/src/config.rs:
```diff
@@ -116,20 +148,20 @@ impl Config {
 
     /// Set the session name
-    pub fn with_session_name(self, name: impl Into<String>) -> Self {
+    pub fn session_name(self, name: impl Into<String>) -> Self {
```
I feel the name `with_session_name` and all the other functions `with` were more meaningful.

didier-wenzek replied:
I agree with @PradeepKiruvale  and I would revert https://github.com/thin-edge/thin-edge.io/pull/1864/commits/bf97c51373fb6fe69dcb7a61a7e90261c9d35cb5 for 3 reasons:

1. This change is unrelated to that PR which is already quite large and introduces critical updates.
2. It's better to have an independent PR for such renaming as this gives people an opportunity to discuss.
3. I'm not sure that adhering to these naming conventions is the highest priority. There are unfortunately worse mistakes to fix first.

I see a benefit to that change, though. This would help the transition to [#[derive(TypedBuilder)]](https://lib.rs/crates/typed-builder) here and many places we have builders.

Bravo555 replied:
> 1. This change is unrelated to that PR which is already quite large and introduces critical updates.
> 2. It's better to have an independent PR for such renaming as this gives people an opportunity to discuss.

I agree, these reasons alone are very convincing, as 60 files changed is indeed a relatively large PR. I'll remove this commit and introduce it in some other refactoring-focused PR in the future, so we can discuss it there.

Bravo555 commented:
I realised that I haven't updated any markdown documentation, but it would probably be good to inform users about this new functionality there. As such, one more commit with documentation will likely be coming.

didier-wenzek commented on configuration/debian/tedge/postinst:
```diff
@@ -3,7 +3,18 @@ set -e
 ### Add include to mosquitto.conf so tedge specific conf will be loaded
 if ! grep -q "/etc/tedge/mosquitto-conf" "/etc/mosquitto/mosquitto.conf"; then
-    echo "include_dir /etc/tedge/mosquitto-conf" >>/etc/mosquitto/mosquitto.conf
+    # Append `include_dir /etc/tedge/mosquitto-conf` before `include_dir
+    # /etc/mosquitto/conf.d` so that conf files in conf.d inherit
+    # `per_listener_settings` defined in /etc/tedge/mosquitto-conf.
+    # `per_listener_settings` has to be defined once, before other listener
+    # settings or else it causes the following error:
+    #
+    # Error: per_listener_settings must be set before any other security
+    # settings.
+    mosquitto_conf=$(awk '/include_dir \/etc\/mosquitto\/conf.d/ \
+    { print "include_dir /etc/tedge/mosquitto-conf"; }1' \
+    /etc/mosquitto/mosquitto.conf)
+    echo "$mosquitto_conf" > /etc/mosquitto/mosquitto.conf
```
As this is a fix to https://github.com/thin-edge/thin-edge.io/issues/1852, one might have to consider to release this independently.

Bravo555 replied:
Sure, I'll submit this snippet in a different PR

Bravo555 replied:
#1919

didier-wenzek replied:
Thanks. Merged.

didier-wenzek commented on docs/src/howto-guides/029_mqtt_local_broker_authentication.md:
```diff
@@ -0,0 +1,166 @@
+# Use MQTT authentication for local broker
+
+thin-edge.io supports using certificate-based authentication when communicating
+with a local broker. There are 3 ways thin-edge.io can use when connecting to
+Mosquitto broker:
```
```suggestion
thin-edge.io supports certificate-based authentication when communicating
with an MQTT broker. Three levels of security are supported:
```

Bravo555 replied:
fixed

didier-wenzek commented on docs/src/howto-guides/029_mqtt_local_broker_authentication.md:
````diff
@@ -0,0 +1,166 @@
+Create a file in `/etc/mosquitto/conf.d/` with the following content and restart
+mosquitto service:
+
+```conf
+listener 8883
+certfile PATH_TO_SERVER_CERTIFICATE
+keyfile  PATH_TO_SERVER_PRIVATE_KEY
+```
+
+- `listener 8883`: defines a new listener on port 8883
+- `certfile`: points to a certificate that will be used by the broker to
+  authenticate itself to connecting clients
+- `keyfile`: points to a private key of the specified certificate, necessary for
+  encrypted communication
````
What is document here is correct. However, I would like to stress that preparing such a configuration file is the point of the `mqtt.external.*` settings of `tedge config`.

```
tedge config list --doc | grep mqtt.external    
mqtt.external.port             Mqtt broker port, which is used by the external mqtt clients to publish or subscribe. Example: 8883
mqtt.external.bind_address     IP address / hostname, which the mqtt broker limits incoming connections on. Example: 0.0.0.0
mqtt.external.bind_interface   Name of network interface, which the mqtt broker limits incoming connections on. Example: wlan0
mqtt.external.capath           Path to a file containing the PEM encoded CA certificates that are trusted when checking incoming client certificates. Example: /etc/ssl/certsNote: If the capath is not set, then no certificates are required for the external connections.
mqtt.external.certfile         Path to the certificate file, which is used by external MQTT listenerExample: /etc/tedge/device-certs/tedge-certificate.pemNote: This setting shall be used together with `mqtt.external.keyfile` for external connections.
mqtt.external.keyfile          Path to the private key file, which is used by external MQTT listenerExample: /etc/tedge/device-certs/tedge-private-key.pemNote: This setting shall be used together with `mqtt.external.certfile` for external connections.
```

- See https://thin-edge.github.io/thin-edge.io/html/howto-guides/013_connect_external_device.html.
- @Bravo555  Change nothing for your current PR. In a follow-up task, we will have to assess if something need to be change for this `mqtt.external` setting.
- @reubenmiller I think this `mqtt.external` setting is badly named making things confusing.

reubenmiller replied:
> @reubenmiller I think this mqtt.external setting is badly named making things confusing.

Yes I fully agree. It is really just a binding address, it does not matter if it is "internal" or "external"...but yes it can be done as a later stage.

didier-wenzek commented on docs/src/howto-guides/029_mqtt_local_broker_authentication.md:
```diff
@@ -0,0 +1,166 @@
+
+## Generating certificates
+
+You can use the following script to generate all required certificates:
```
@reubenmiller We can consider to extend the `tedge cert create` command to create such certificates.

didier-wenzek reviewed (approved):
Nicely and thoroughly done! Thank you! Approved with some comments.

> **settings naming**: there's `mqtt.client.ca_file` for server authentication, but `mqtt.client.auth.cert_file/key_file` for client authentication. All of them are related to authentication, so I think perhaps all of them should be under `mqtt.client.auth`. Additionally, for `cert_file`, `key_file` we could remove `_file` suffix, at for anybody familiar with authentication, it's evident that we only provide one cert-private key pair. For `ca_file`, we could perhaps do something else to eliminate the underscore.

- I would keep the same names as mosquitto e.g `certfile` and `keyfile`.
- We also need to be consistent with the misc `mqtt.external.*` settings.

> **config organisation**: because of current config architecture, I had to include some additional boilerplate for checking invariants like whether or not if client certificate was included, was the private key also included and vice-versa. As described [here](https://github.com/thin-edge/thin-edge.io/issues/1812#issuecomment-1486882503) I think this aspect of validation configuration states should be moved from config clients to config provider.

Indeed there is some boilerplate, but this okay for now.

> **RobotFramework tests**: When I started working on this PR, I did a small RobotFramework suite `Tests.Mqtt.Basic Pub Sub` which verified all the MQTT clients work correctly with each: no authentication, server authentication, server + client authentication. Then, because we ideally want to test all clients, I've edited the test suite so all the tests work with server +client authentication, testing the largest possible surface. Now that all the tests work with authentication, should the original  `Tests.Mqtt.Basic Pub Sub` suite be removed? It's somewhat trivial and things it tests are already tested by the rest of the tests.

Are you speaking about [these tests](https://github.com/thin-edge/thin-edge.io/pull/1864/files#diff-11abc9ff9cf4faad9523fea5f7fd79ab823e05567635e4b40f9dda0bb6c044a9)?

I would keep them even if redundant. Indeed there are simple and cover well the 3 levels of authentication.

Bravo555 commented on crates/core/tedge/src/cli/mqtt/cli.rs:
```diff
@@ -61,6 +62,25 @@ impl BuildCommand for TEdgeMqttCli {
             .load()?
             .query(MqttClientCapathSetting)
             .ok();
+        let client_cert = context
+            .config_repository
+            .load()?
+            .query(MqttClientAuthCertSetting);
+        let client_private_key = context
+            .config_repository
+            .load()?
+            .query(MqttClientAuthKeySetting);
```
`tedge` does not depend on `mqtt_channel` so I felt it'd be better not to introduce a dependency to save a few lines.

Bravo555 commented on crates/common/mqtt_channel/src/config.rs:
```diff
@@ -188,44 +232,67 @@ impl Config {
 
     /// Wrap this config into an internal set of options for `rumqttc`.
-    pub(crate) fn mqtt_options(&self) -> rumqttc::MqttOptions {
+    pub fn mqtt_options(&self) -> Result<rumqttc::MqttOptions, rustls::Error> {
```
Fixed in fixup

Bravo555 commented on crates/core/tedge_mapper/src/c8y/mapper.rs:
```diff
@@ -87,19 +88,17 @@ impl TEdgeComponent for CumulocityMapper {
             &service_type,
+            Some(&tedge_config),
             mqtt_host.clone(),
             mqtt_port,
```
fixed

Bravo555 commented on crates/core/tedge_mapper/src/c8y/mapper.rs:
```diff
@@ -186,12 +183,19 @@ pub async fn create_mqtt_client_will_message(
     service_type: &str,
+    tedge_config: Option<&TEdgeConfig>,
     mqtt_host: String,
     mqtt_port: u16,
```
fixed

Bravo555 commented:
Added fixups addressing review comments.

Also renamed `mqtt.client.ca_file/dir` to `mqtt.client.auth.cafile/dir` so options related to authentication are all grouped together.

didier-wenzek commented:
Thank you for the fixups.

Please, rebase and squash the fixup commits. And I will merge this PR.

Bravo555 commented:
@didier-wenzek squashed

didier-wenzek commented:
> @didier-wenzek squashed

Unfortunately, two system tests are failing on a certificate issue.

Cargo audit, is also failing: did you rebase once your fix merged?

Bravo555 commented:
> Cargo audit is also failing: did you rebase once your fix merged?
My bad, forgot to rebase, will fix
